fvm/errors/codes.go

```diff
@@ -4,26 +4,27 @@ import "fmt"
 
 type ErrorCode uint16
 
-func (ec ErrorCode) String() string {
-	return fmt.Sprintf("[Error Code: %d]", ec)
+func (ec ErrorCode) IsFailure() bool {
+	return ec >= FailureCodeUnknownFailure
```

Comment:
It's cool idea that we combine both error type and reuse code.

I had a different solution: since failures are fatal errors, instead of making a Failure type, we could just use fmt.Errorf to create them. 

For instance: 
```
// NewEncodingFailuref formats and returns a new EncodingFailure
func NewEncodingFailuref(msg string, err error) error {
        // Note we don't want to use %w to wrap the error, because that might cause the errors.As to return true 
        // when checking if the error is non-fatal sentinel error.
	return fmt.Errorf("%v, error: %v", msg, err) 
}
```

Based on [our error handling guideline](https://github.com/onflow/flow-go/blob/master/CodingConventions.md#best-practice-guidelines), any non-sentinal errors are fatal errors, so we don't have to define types for them. We can still keep the error code, they can be included in the error message, and get printed out into logs.

Reply:
I agree that we can ignore the error code altogether if the error is not user facing, and if we never special case error handling (which we clearly do with the SplitError stuff).

right now, both fvm and flow transactions users are stuck with checking the error string, rather than checking the error code if we (or the user) want to handle specific types of error.  Dropping the error code (/ wrapping fmt.Errorf) doesn't help in that regard.

Reply:
We have plan to get rid of the `SplitErrorType`. It will go away when getting rid of the `Failure` type. 

>right now, both fvm and flow transactions users are stuck with checking the error string, rather than checking the error code if we (or the user) want to handle specific types of error. Dropping the error code (/ wrapping fmt.Errorf) doesn't help in that regard.

I don't understand the difference. 

flow transaction users will only see non-fatal error, probably won't see fatal errors, we will include the first code as the code in the error message, so that for instance, if a user hits fail to check storage capacity due to interaction limit hit error, the error message will include error code for interaction limit hit error. And whether the error code is included in the error message or in a separate field of the response is a separate question, right? 

for fvm, when they see fatal errors, they probably don't care about the error code, because no matter what code, it will crash the component anyway. And when they see non-fatal errors, since all errors are wrapped, they just need to unwrap each layer in order to handle a specific error and get that error code. In other words, they can check the error code instead of checking the error string. right?

Reply:
I think lets (eventually) get rid of the failure codes (they will never be seen by the user) and failures as a concept, and just have everything that is not a coded error (or does not unwrap to one) to be a failure. 

Since we currently have location where we indiscriminately wrap errors into coded errors and then later try to unwrap them as a failure first, this change is not something we can easily do.